examples/benchmark/main.rs

```diff
@@ -0,0 +1,271 @@
+                        "vm:balancer_v2",
+                        tvl_filter.clone(),
+                        Some(balancer_pool_filter),
+                    )
+                    .auth_key(Some(tycho_api_key.clone()))
+                    .set_tokens(all_tokens.clone())
+                    .await
+                    .build()
+                    .await
+                    .expect("Failed building Balancer V2 protocol stream")
+                    .boxed(),
+                "curve" => ProtocolStreamBuilder::new(&tycho_url, Chain::Ethereum)
+                    .exchange::<EVMPoolState<PreCachedDB>>(
+                        "vm:curve",
+                        tvl_filter.clone(),
+                        Some(curve_pool_filter),
+                    )
+                    .auth_key(Some(tycho_api_key.clone()))
+                    .set_tokens(all_tokens.clone())
+                    .await
+                    .build()
+                    .await
+                    .expect("Failed building Curve protocol stream")
+                    .boxed(),
+                _ => {
+                    eprintln!("Unknown protocol: {}", protocol);
+                    continue;
+                }
+            };
+
+            info!("BENCHMARKING {} protocol on {} swaps", protocol, cli.n_swaps);
+            let times = benchmark_swaps(stream, cli.n_swaps).await;
+            results.insert(protocol, times);
+        }
+        // Add a small delay to ensure the WebSocket disconnection completes
+        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+    }
+
+    analyze_results(&results, cli.n_swaps);
+}
+
+async fn benchmark_swaps(
+    mut protocol_stream: BoxStream<'_, Result<BlockUpdate, StreamDecodeError>>,
+    n: usize,
+) -> Vec<u128> {
+    let mut times = Vec::new();
+    let mut pairs: HashMap<String, Vec<Token>> = HashMap::new();
+
+    if let Some(Ok(message)) = protocol_stream.next().await {
+        for (id, comp) in message.new_pairs.iter() {
+            pairs
+                .entry(id.clone())
+                .or_insert_with(|| comp.tokens.clone());
+        }
+
+        if message.states.is_empty() {
+            return times;
+        }
+
+        for (id, tokens) in pairs.iter().cycle() {
+            if let Some(state) = message.states.get(id) {
+                let amount_in =
+                    BigUint::from(1u32) * BigUint::from(10u32).pow(tokens[0].decimals as u32);
+
+                let start = Instant::now();
+                let _ = state.get_amount_out(amount_in.clone(), &tokens[0], &tokens[1]);
+                let duration = start.elapsed().as_nanos();
+
+                times.push(duration);
+
+                info!("Swap {} -> {} took {} ns", tokens[0].symbol, tokens[1].symbol, duration);
+
+                if times.len() >= n {
+                    break;
+                }
+            }
+        }
+    }
+
+    times
+}
+
+fn calculate_std_dev(times: &[u128], avg: f64) -> f64 {
```

Comment:
can we move these two functions down? calculate_std_dev and calculate_trimmed_mean. so that it is in reading order

Reply:
Sure!